### ISSUE TITLE
Scalable applications without load balancer

### ISSUE THREAD

**Issue by tomhughesnice** [boxfuse/boxfuse-issues#47]
Is it possible to create scalable applications but without the Amazon Elastic Load balancer?
When all my microservice applications launch they register with a self hosted DNS(https://github.com/Netflix/eureka). This handles the routing of requests to app instances and the Elastic load balancer is not hit at all.

**Comment by axelfontaine** [comment]
This would imply there would be no URL/domain mapping for them either. This also affects health checks, which would then have to be performed by Boxfuse on ASG creation, instead of having the ELB take care of this.

In a nutshell if we have version 1 running and you deploy version 2 with Boxfuse you would now have:
1. Boxfuse creates the AMI as well as the ASG for v2
2. Boxfuse performs the healthcheck on the initial instances of the ASG
3. Boxfuse terminates the v1 ASG

Does this make sense? Or did I miss something?

**Comment by axelfontaine** [comment]
Hi Tom,

what are your thoughts on this?

Best regards,
Axel

**Comment by tomhughesnice** [comment]
Hi Axel, 
Haven't been looking at this for a while and wanted to touch on it again.

In response to your post from November 26th, no URL/domain mapping would be required. The service running on this instance would register with the eureka(Service discovery). Eureka would hold the url mapping for each service which other services could query.

I would like one or two services to have an Amazon Elastic Load balancer as they will be client facing for websites, mobile apps and the like. But the rest of my services will not require the Elastic load balancer as they will not be client facing.

` a nutshell if we have version 1 running and you deploy version 2 with Boxfuse you would now have:
1. Boxfuse creates the AMI as well as the ASG for v2
2. Boxfuse performs the healthcheck on the initial instances of the ASG
3. Boxfuse terminates the v1 ASG`

Your deployment scenario makes sense to me as long as Boxfuse creates an identical number of instances to V1. Perhaps we have scaled up to 5 instances for example in a period of heavy load.

In my current deployment I have 12 Elastic load balancers created across my production and test environment. In reality I only need 2 of these, the rest are redundant and never used.

**Comment by axelfontaine** [comment]
Thanks for the update Tom.

(Unrelated but important: could you get back to me by email about the VAT issue I contacted you about?)

**Comment by tomhughesnice** [comment]
Brilliant, thats good to know.

I would assume implementation for this feature would be performed at the same time as #41 . Would you have an eta on the rollout?

Once put together I could help you put together an archetype for a Spring cloud microservice application built using boxfuse. Very hard to find a documented method to test and deploy microservices to AWS. Could make this whole process much easier for developers.

**Comment by axelfontaine** [comment]
An archetype would be awesome!

This is actually reasonably high priority for us as we have numerous customers that have been asking for this. This could also be a nice workaround for #48 until we release proper SQS & SNS support.

You can expect this before the end of April.

**Comment by tomhughesnice** [comment]
Any update on this @axelfontaine?

**Comment by axelfontaine** [comment]
Thanks for checking in Tom. April was incredibly busy for us, however I can assure you this is very high on our priority list. I'll keep you updated.

**Comment by tomhughesnice** [comment]
Ok if you could let me know that would be sweet. Save me checking the Release notes on you website every day.

**Comment by axelfontaine** [comment]
This is our current proposal for the initial implementation of this feature:

We would introduce a new app type called `worker`. (We may introduce more specialized ones for SQS, SNS and other inputs later)

The lifecycle of a worker app would be like this:

**Launching v1**

1. You would now have 3 new optional image configuration attributes called `notification.port` (the port Boxfuse should use to notify your instances of lifecycle events), `notification.path` (the path Boxfuse should use to notify your instances of lifecycle events) & `notification.timeout` (the time in seconds Boxfuse will wait to act upon a notification (see kill description below))
2. Boxfuse creates the image, uploads it to the Vault and converts it to an AMI as usual
3. Boxfuse launches an auto-scaling group (fixed-capacity for now, auto-scaling will be added later) based on that AMI
4. Each instance also receives an environment variable called `BOXFUSE_NOTIFICATION_TOKEN`
5. Boxfuse performs a healthcheck on each instance
6. When all instances are healthy, Boxfuse sends an HTTP POST with a notification payload to each instance at the configured notification port & path (if those have been configured) with the following contents:

```
{
"source":"boxfuse",
"event":"in-service",
"time":"2016-05-07T12:34:56Z",
"token":"abc123"
}
```

where `time` is the current time in UTC and `token` is the same string value as the one contained in the `BOXFUSE_NOTIFICATION_TOKEN` environment variable of your instance (allowing you to validate the request).

**Launching v2**

1. Steps 1-6 identical to v1 (both v1 and v2 are now up and running)
2. Boxfuse sends a notification payload to the v1 instances at the configured notification port & path (if those have been configured) with the following contents:
```
{
"source":"boxfuse",
"event":"kill",
"time":"2016-05-07T13:34:56Z",
"token":"abc123"
}
```
where `time` is the current time in UTC + the `notification.timeout` and `token` is the same string value as the one contained in the `BOXFUSE_NOTIFICATION_TOKEN` environment variable of your instance (allowing you to validate the request).
3. Boxfuse kills the v1 ASG at the time specified in the notification payload (leaving only the v2 ASG and its instances up and running)

@tomhughesnice @herder What are your thoughts on this? Does this look good to you or is there something critical we should adjust before we move forward with this?

**Comment by herder** [comment]
This seems like a reasonable solution to me. 👍

**Comment by tomhughesnice** [comment]
Seems like a good solution to handle the lifecycle of a worker app.
Lack of scaling is the real shortcoming though.

**Comment by axelfontaine** [comment]
This has now been implemented. There is a new application type called `worker` which is for auto-scaled applications without an ELB (think Eureka and client-side load-balancing). All current auto-scaling features of load-balanced applications are supported.

In addition we have also made some changes to VirtualBox networking to better support this for local development. See https://github.com/boxfuse/boxfuse-issues/issues/84#issuecomment-237170384

We have deferred the implementation of the notifications described above, as we are currently tending towards using standard auto-scaling group+SNS notifications. We are planning to revisit this as part of #103 with dedicated support for long-running workers behind queues.
